### PR TITLE
lib: introduce imap0, imap1

### PR DESCRIPTION
**Pull request by zimbatm**
###### Motivation for this change

For historical reasons, imap starts counting at 1 and it's not
consistent with the rest of the lib.

So for now we split imap into imap0 that starts counting at zero and
imap1 that starts counting at 1. And imap is marked as deprecated.

See https://github.com/NixOS/nixpkgs/commit/c71e2d42359f9900ea2c290d141c0d606471da16#commitcomment-21873221

Also:
```
$ ag "imap " | wc -l
115
```

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by zimbatm** [comment]
If I get :+1: for this PR I will also start replacing all the imap occurences

**Comment by Profpatsch** [comment]
I thought about maybe having `imapN n f xs`, but that’s probably a bad idea because that’s trivially just `n: f: imap0 (f n)` and complicates the function.
The original `imap` name is probably gone for good, because of the semantics change (without a type change). So yeah, +1 from me.

**Comment by zimbatm** [comment]
In lib/strings.nix we also have concatImapStrings and concatImapStringsSep that start counting at 1

**Comment by zimbatm** [comment]
After replacing all the lib.imap calls there is only one instance where imap0 was used.

**Review by 0xABAB — APPROVED** [review]

File: lib/lists.nix [hunk 1]
```diff
@@ -77,15 +77,30 @@ rec {
...
+  imap1 = f: list: genList (n: f (n + 1) (elemAt list n)) (length list);
+
+  /* deprecated:
+
+     For historical reasons, imap has an index starting at 1.
+
+     But for consistency with the rest of the library we want an index
+     starting at zero.
+  */
+  imap = imap1;
```
Comment (nbp) [inline lib/lists.nix]: nit: move it to `lib/deprecated.nix`

**Review by nbp — APPROVED** [review]

File: lib/lists.nix (hunk 1 above)
Comment (zimbatm) [inline lib/lists.nix]: fixed in 54a81c88aaa524a4b5de26870f81074ea247451d

**Comment by NeQuissimus** [comment]
@mguentner @zimbatm I fixed booter.nix in d788244423 since I was just building something off master and got lots of `error: attribute ‘imap’ missing, at /home/nequi/dev/nixpkgs/pkgs/stdenv/booter.nix:76:30` :)

**Comment by zimbatm** [comment]
thanks @NeQuissimus

**Pull request merged**